File: packages/python/src/seclab_suite_runtime/operation_logs.py
"""语义操作事件构建、序列化和校验。"""

import re
import secrets
import time
import uuid
from collections.abc import Mapping
from dataclasses import dataclass, field

from .errors import InvalidEvent
from .types import OperationImpact, OperationOutcome, OperationTarget, ParameterValue

EVENT_CODE = re.compile(r"^[a-z][a-z0-9_]{2,63}$")
SENSITIVE_KEYS = (
    "password",
    "token",
    "authorization",
    "secret",
    "cookie",
    "command",
    "environment",
)
OPERATION_CONTEXT_HEADER = "x-seclab-operation-context"


def operation_context_from_headers(headers: Mapping[str, str]) -> str | None:
    """从套件代理请求头提取可信操作上下文 ID。"""
    value = headers.get(OPERATION_CONTEXT_HEADER)
    if value is None:
        return None
    value = value.strip()
    return value if value and len(value) <= 128 else None


def uuid7() -> uuid.UUID:
    """生成符合 RFC 9562 布局的 UUIDv7。"""
    timestamp = int(time.time() * 1000) & ((1 << 48) - 1)
    value = timestamp << 80
    value |= 0x7 << 76
    value |= secrets.randbits(12) << 64
    value |= 0b10 << 62
    value |= secrets.randbits(62)
    return uuid.UUID(int=value)


@dataclass
class OperationEvent:
    event_code: str
    zh_cn: str
    en_us: str
    outcome: OperationOutcome
    impact: OperationImpact
    event_id: str = field(default_factory=lambda: str(uuid7()))
    operation_context_id: str | None = None
    target: OperationTarget | None = None
    task_id: str | None = None
    parameters: dict[str, ParameterValue] = field(default_factory=dict)
    error_code: str | None = None
    error_summary: str | None = None

    def validate(self) -> None:
        try:
            identifier = uuid.UUID(self.event_id)
        except ValueError as error:
            raise InvalidEvent("eventId must be a UUIDv7") from error
        if identifier.version != 7:
            raise InvalidEvent("eventId must be a UUIDv7")
        if self.operation_context_id is not None and (
            not self.operation_context_id.strip() or len(self.operation_context_id) > 128
        ):
            raise InvalidEvent("operationContextId must contain 1 to 128 characters")
        if not EVENT_CODE.fullmatch(self.event_code):
            raise InvalidEvent("eventCode must use lower snake case")
        if any(not value.strip() or len(value) > 128 for value in (self.zh_cn, self.en_us)):
            raise InvalidEvent("event labels must contain 1 to 128 characters")
        if len(self.parameters) > 32 or any(
            marker in key.lower() for key in self.parameters for marker in SENSITIVE_KEYS
        ):
            raise InvalidEvent("operation parameters contain unsupported fields")

    def to_dict(self) -> dict[str, object]:
        self.validate()
        target: dict[str, object] | None = None
        if self.target is not None:
            target = {
                "kind": self.target.kind,
                "id": self.target.id,
                "displayName": self.target.display_name,
                "ownership": self.target.ownership,
            }
        return {
            "eventId": self.event_id,
            "operationContextId": self.operation_context_id,
            "eventCode": self.event_code,
            "eventLabel": {"zhCn": self.zh_cn, "enUs": self.en_us},
            "outcome": self.outcome.value,
            "impact": self.impact.value,
            "target": target,
            "taskId": self.task_id,
            "parameters": self.parameters,
            "errorCode": self.error_code[:128] if self.error_code else None,
            "errorSummary": redact_error(self.error_summary) if self.error_summary else None,
        }


def redact_error(value: str) -> str:
    sanitized = value.replace("\r", " ").replace("\n", " ")
    lowered = sanitized.lower()
    positions = [
        lowered.find(marker)
        for marker in ("bearer ", "token=", "password=")
        if lowered.find(marker) >= 0
    ]
    if positions:
        sanitized = sanitized[: min(positions)] + "[REDACTED]"
    return sanitized[:512]

File: packages/rust/src/operation_logs.rs
//! 套件语义操作事件构建与校验。

use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use uuid::Uuid;

use crate::{Error, OperationImpact, OperationOutcome, OperationTarget, ParameterValue, Result};

pub const OPERATION_CONTEXT_HEADER: &str = "x-seclab-operation-context";

/// 从代理请求头值提取可信操作上下文 ID。
pub fn operation_context_from_header(value: Option<&str>) -> Option<String> {
    value
        .map(str::trim)
        .filter(|value| !value.is_empty() && value.len() <= 128)
        .map(ToOwned::to_owned)
}

/// 动态事件双语名称。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationEventLabel {
    pub zh_cn: String,
    pub en_us: String,
}

/// 套件提交给 Agent 的操作事件。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OperationEvent {
    pub event_id: String,
    pub operation_context_id: Option<String>,
    pub event_code: String,
    pub event_label: OperationEventLabel,
    pub outcome: OperationOutcome,
    pub impact: OperationImpact,
    pub target: Option<OperationTarget>,
    pub task_id: Option<String>,
    pub parameters: BTreeMap<String, ParameterValue>,
    pub error_code: Option<String>,
    pub error_summary: Option<String>,
}

/// 逐步构建并在提交前校验事件。
pub struct OperationEventBuilder(OperationEvent);

impl OperationEvent {
    /// 创建一个带 UUIDv7 幂等键的事件构建器。
    pub fn builder(
        event_code: impl Into<String>,
        zh_cn: impl Into<String>,
        en_us: impl Into<String>,
        outcome: OperationOutcome,
        impact: OperationImpact,
    ) -> OperationEventBuilder {
        OperationEventBuilder(Self {
            event_id: Uuid::now_v7().to_string(),
            operation_context_id: None,
            event_code: event_code.into(),
            event_label: OperationEventLabel {
                zh_cn: zh_cn.into(),
                en_us: en_us.into(),
            },
            outcome,
            impact,
            target: None,
            task_id: None,
            parameters: BTreeMap::new(),
            error_code: None,
            error_summary: None,
        })
    }

    /// 校验 Agent 接口公开约束。
    pub fn validate(&self) -> Result<()> {
        let id = Uuid::parse_str(&self.event_id)
            .map_err(|_| Error::InvalidEvent("eventId must be a UUIDv7".to_string()))?;
        if id.get_version_num() != 7 {
            return Err(Error::InvalidEvent("eventId must be a UUIDv7".to_string()));
        }
        if self
            .operation_context_id
            .as_deref()
            .is_some_and(|value| value.trim().is_empty() || value.len() > 128)
        {
            return Err(Error::InvalidEvent(
                "operationContextId must contain 1 to 128 characters".to_string(),
            ));
        }
        let code = self.event_code.as_bytes();
        if !(3..=64).contains(&code.len())
            || !code[0].is_ascii_lowercase()
            || !code
                .iter()
                .all(|value| value.is_ascii_lowercase() || value.is_ascii_digit() || *value == b'_')
        {
            return Err(Error::InvalidEvent(
                "eventCode must use lower snake case".to_string(),
            ));
        }
        for label in [&self.event_label.zh_cn, &self.event_label.en_us] {
            if label.trim().is_empty() || label.chars().count() > 128 {
                return Err(Error::InvalidEvent(
                    "event labels must contain 1 to 128 characters".to_string(),
                ));
            }
        }
        if self.parameters.len() > 32 || self.parameters.keys().any(|key| sensitive_key(key)) {
            return Err(Error::InvalidEvent(
                "operation parameters contain unsupported fields".to_string(),
            ));
        }
        Ok(())
    }
}

impl OperationEventBuilder {
    pub fn operation_context_id(mut self, value: impl Into<String>) -> Self {
        self.0.operation_context_id = Some(value.into());
        self
    }

    pub fn target(mut self, target: OperationTarget) -> Self {
        self.0.target = Some(target);
        self
    }

    pub fn task_id(mut self, value: impl Into<String>) -> Self {
        self.0.task_id = Some(value.into());
        self
    }

    pub fn parameter(mut self, key: impl Into<String>, value: ParameterValue) -> Self {
        self.0.parameters.insert(key.into(), value);
        self
    }

    pub fn error(mut self, code: impl Into<String>, summary: impl AsRef<str>) -> Self {
        self.0.error_code = Some(code.into().chars().take(128).collect());
        self.0.error_summary = Some(redact_error(summary.as_ref()));
        self
    }

    pub fn build(self) -> Result<OperationEvent> {
        self.0.validate()?;
        Ok(self.0)
    }
}

fn sensitive_key(key: &str) -> bool {
    let key = key.to_ascii_lowercase();
    [
        "password",
        "token",
        "authorization",
        "secret",
        "cookie",
        "command",
        "environment",
    ]
    .iter()
    .any(|value| key.contains(value))
}

fn redact_error(value: &str) -> String {
    let mut value = value.replace(['\r', '\n'], " ");
    for marker in ["Bearer ", "token=", "password="] {
        if let Some(index) = value
            .to_ascii_lowercase()
            .find(&marker.to_ascii_lowercase())
        {
            value.truncate(index);
            value.push_str("[REDACTED]");
        }
    }
    value.chars().take(512).collect()
}
